internal/git/util.go

```diff
@@ -245,14 +268,177 @@ func cloneRepo(ctx context.Context, gitSync *v1alpha1.GitSync, options *git.Clon
 	path := getLocalRepoPath(gitSync)
 
 	r, err := git.PlainCloneContext(ctx, path, false, options)
-	if err != nil && errors.Is(err, git.ErrRepositoryAlreadyExists) {
-		// open the existing repo and return it.
-		existingRepo, openErr := git.PlainOpen(path)
-		if openErr != nil {
-			return r, fmt.Errorf("failed to open existing repo, err: %v", openErr)
+	if err != nil {
+		if errors.Is(err, git.ErrRepositoryAlreadyExists) {
+			// Open the existing repo and return it.
+			existingRepo, openErr := git.PlainOpen(path)
+			if openErr != nil {
+				return nil, fmt.Errorf("failed to open existing repo, err: %v", openErr)
+			}
+			return existingRepo, nil
+		}
+		return nil, err
+	}
+
+	// No need to fetch the references if  the target revision is already main or master (branches)
+	if gitSync.Spec.TargetRevision != "main" && gitSync.Spec.TargetRevision != "master" {
+		// fetch all references
+		err = fetchAll(r)
```

Comment:
why we need to fetch all brach of repo if we are only dealing with single branch/tag, as this will run in each sync which might cause performance issue if repo having more than 100 branches.

Reply:
I believe we can use the `targetReference` from gitsync spec to clone the particular branch/tag ? @xdevxy

Reply:
we have to fetch such that we can checkout to those branch or tag ,by default only the default reference is available and then if i do checkout it gives error detached head. Recursive process can be avoided by putting the  check if the repo is already checked to the  targetRevision branch or tag.